Error parsing rust ternery expressions of if else

Preparing to parse files. 
Attempting to parse file: test.rs
beginning null denotation identifier handling
current peek token when starting path pratt parse: '>'
current token (just about to return path to null denotation): '>'
finished null denotation identifier path parsing - next is branching 
test.rs:12:13: error: found unexpected token ‘if’ in null denotation
   12 |     let c = if a > 10 { true } else { false };
      |             ^
null denotation is null; returning null for parse_expr
test.rs:12:16: error: failed to parse expression in let statement
   12 |     let c = if a > 10 { true } else { false };
      |                ^
test.rs:12:5: error: failed to parse statement or expression without block in block expression
   12 |     let c = if a > 10 { true } else { false };
      |     ^
test.rs:13:1: error: unrecognised token ‘}’ for start of item
   13 | }
      | ^
test.rs:13:1: error: failed to parse item in crate


```
fn main() {
    let a = 123;

    let b = {
        if a > 10 {
            true
        } else {
            false
        }
    };

    let c = if a > 10 { true } else { false };
}
```

Here the parser fails to parse the assignment to c.

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -7504,12 +7504,23 @@ Parser<ManagedTokenSource>::parse_loop_label ()
 template <typename ManagedTokenSource>
 std::unique_ptr<AST::IfExpr>
 Parser<ManagedTokenSource>::parse_if_expr (
-  std::vector<AST::Attribute> outer_attrs ATTRIBUTE_UNUSED)
+  std::vector<AST::Attribute> outer_attrs ATTRIBUTE_UNUSED, bool pratt_parse)
 {
   // TODO: make having outer attributes an error?
-
-  Location locus = lexer.peek_token ()->get_locus ();
-  skip_token (IF);
+  Location locus = Linemap::unknown_location ();
+  if (!pratt_parse)
+    {
+      locus = lexer.peek_token ()->get_locus ();
+      if (!skip_token (IF))
+	{
+	  skip_after_end_block ();
+	  return nullptr;
+	}
+    }
+  else
+    {
+      locus = lexer.peek_token ()->get_locus () - 1;
+    }
 
   // detect accidental if let
   if (lexer.peek_token ()->get_id () == LET)
@@ -7640,12 +7651,23 @@ Parser<ManagedTokenSource>::parse_if_expr (
 template <typename ManagedTokenSource>
 std::unique_ptr<AST::IfLetExpr>
 Parser<ManagedTokenSource>::parse_if_let_expr (
-  std::vector<AST::Attribute> outer_attrs ATTRIBUTE_UNUSED)
+  std::vector<AST::Attribute> outer_attrs ATTRIBUTE_UNUSED, bool pratt_parse)
 {
   // TODO: make having outer attributes an error?
-
-  Location locus = lexer.peek_token ()->get_locus ();
-  skip_token (IF);
+  Location locus = Linemap::unknown_location ();
+  if (!pratt_parse)
+    {
+      locus = lexer.peek_token ()->get_locus ();
+      if (!skip_token (IF))
+	{
+	  skip_after_end_block ();
+	  return nullptr;
+	}
+    }
+  else
+    {
+      locus = lexer.peek_token ()->get_locus () - 1;
+    }
 
   // detect accidental if expr parsed as if let expr
   if (lexer.peek_token ()->get_id () != LET)
@@ -12399,6 +12421,18 @@ Parser<ManagedTokenSource>::null_denotation (
     case LEFT_CURLY:
       // ok - this is an expression with block for once.
       return parse_block_expr (std::move (outer_attrs), true);
+    case IF:
+      // if or if let, so more lookahead to find out
+      if (lexer.peek_token (1)->get_id () == LET)
+	{
+	  // if let expr
+	  return parse_if_let_expr (std::move (outer_attrs), true);
+	}
+      else
+	{
+	  // if expr
+	  return parse_if_expr (std::move (outer_attrs), true);
+	}
     case MATCH_TOK:
       // also an expression with block
       return parse_match_expr (std::move (outer_attrs), true);

--- a/gcc/rust/parse/rust-parse.h
+++ b/gcc/rust/parse/rust-parse.h
@@ -480,10 +480,12 @@ private:
 		    bool pratt_parse = false);
   std::unique_ptr<AST::IfExpr>
   parse_if_expr (std::vector<AST::Attribute> outer_attrs
-		 = std::vector<AST::Attribute> ());
+		 = std::vector<AST::Attribute> (),
+		 bool pratt_parse = false);
   std::unique_ptr<AST::IfLetExpr>
   parse_if_let_expr (std::vector<AST::Attribute> outer_attrs
-		     = std::vector<AST::Attribute> ());
+		     = std::vector<AST::Attribute> (),
+		     bool pratt_parse = false);
   std::unique_ptr<AST::LoopExpr>
   parse_loop_expr (std::vector<AST::Attribute> outer_attrs
 		   = std::vector<AST::Attribute> (),

--- a/gcc/testsuite/rust.test/compilable/block_expr_parser_bug.rs
+++ b/gcc/testsuite/rust.test/compilable/block_expr_parser_bug.rs
@@ -1,0 +1,4 @@
+fn main() {
+    let a = 123;
+    let b = if a > 10 { a - 1 } else { a + 1 };
+}

